### PR TITLE
[4.x] Nima webclient http/2

### LINKED ISSUE #4910: Nima - HTTP/2 Client
#### Flow control
- [x] Server inbound FC (outbound is already in main)
- [x] Client outbound FC
- [x] Client inbound FC

#### Frame splitting by [MAX_FRAME_SIZE](https://httpwg.org/specs/rfc9113.html#rfc.section.4.2)
- [x] Client 
- [x] Server

### Followup issues

- #6543
- #6544
- #6545
- #6546
- #6547
- #6548

### PR DESCRIPTION
**Pull request by danielkec**
Fixes #4910

### CODE REVIEW

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/Http2ClientConnection.java [hunk 1]
```diff
@@ -74,22 +118,93 @@ Http2ClientConnection connect() {
...
+                        Http2WindowUpdate http2WindowUpdate = Http2WindowUpdate.create(data);
+                        recvListener.frameHeader(helidonSocket, frameHeader);
+                        recvListener.frame(helidonSocket, http2WindowUpdate);
+                        //TODO: flow-control
+                        return null;
+                    }
+                }
+            }
+
+            LOGGER.log(System.Logger.Level.DEBUG, "Deffering frame " + frameHeader.type() + " with streamId:"
```
Comment (zimmi) [inline nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/Http2ClientConnection.java:190]: Random drive by comment: It seems to be wasteful to allocate a new String just to throw it away when DEBUG is not active. Is there a reason to not use the format String variant instead?

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.Logger.html#log(java.lang.System.Logger.Level,java.lang.String,java.lang.Object...)

Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/Http2ClientConnection.java:190]: Still in progress, I am logging this for development purposes, logging will be top-notch before PR is ready for review :slightly_smiling_face:.

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java [hunk 2]
```diff
@@ -217,17 +223,41 @@ private Http2Headers prepareHeaders(WritableHeaders<?> headers) {
...
         if (explicitConnection == null) {
-            ConnectionKey connectionKey = new ConnectionKey(uri.scheme(), uri.authority(), uri.host(), uri.port(), tls);
+            return newStream(uri);
+        } else {
+            throw new UnsupportedOperationException("Explicit connection not (yet) supported for HTTP/2 client");
+        }
+    }
+
+    private Http2ClientStream newStream(UriHelper uri){
+        //todo default by scheme should be somewhere else
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java:233]: Crossed my mind but there was a test for that:
```java
    @Test
    void testDefaults() {
        UriQueryWriteable query = UriQueryWriteable.create();
        UriHelper helper = UriHelper.create(URI.create("http://localhost"), query);

        assertThat(helper.authority(), is("localhost"));
        assertThat(helper.host(), is("localhost"));
        assertThat(helper.path(), is(""));
        assertThat(helper.port(), is(-1));
        assertThat(helper.scheme(), is("http"));
    }
```

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java [hunk 3]
```diff
@@ -95,7 +97,10 @@ public Http2ClientRequest pathParam(String name, String value) {
 
     @Override
     public Http2ClientRequest queryParam(String name, String... values) {
-        throw new UnsupportedOperationException("Not implemented");
+        for (String val : values) {
+            query.add(name, val);
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java:101]: It does not but I can add it

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java [hunk 4]
```diff
@@ -39,6 +39,8 @@
 import io.helidon.nima.webclient.UriHelper;
 
 class ClientRequestImpl implements Http2ClientRequest {
+    static final HeaderValue USER_AGENT_HEADER = Header.create(Header.USER_AGENT, "Helidon Nima WebClient");
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java]: np

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java]: added

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java]: `port()` method modified

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java [hunk 5]
```diff
@@ -217,17 +223,41 @@ private Http2Headers prepareHeaders(WritableHeaders<?> headers) {
...
+        int port = uri.port();
+        if (uri.port() == -1 && "https".equals(uri.scheme())) {
+            port = 443;
+        }
+        if (uri.port() == -1 && "http".equals(uri.scheme())) {
+            port = 80;
+        }
+
+        try {
+            ConnectionKey connectionKey = new ConnectionKey(method,
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java:233]: Seen!

File: nima/http2/webclient/src/test/java/io/helidon/nima/http2/webclient/Http2WebClientTest.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+    private static final Http.HeaderName CLIENT_CUSTOM_HEADER_NAME = Http.Header.create("client-custom-header");
+    private static final Http.HeaderName SERVER_CUSTOM_HEADER_NAME = Http.Header.create("server-custom-header");
+    private static final Http.HeaderName SERVER_HEADER_FROM_PARAM_NAME = Http.Header.create("header-from-param");
+    private static final Http.HeaderName CLIENT_USER_AGENT_HEADER_NAME = Http.Header.create("client-user-agent");
+    private static WebServer server;
+    private static ExecutorService executorService;
+
+    @BeforeAll
+    static void beforeAll() {
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/test/java/io/helidon/nima/http2/webclient/Http2WebClientTest.java]: It does

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java [hunk 7]
```diff
@@ -118,6 +123,7 @@ public Http2ClientResponse submit(Object entity) {
             entityBytes = entityBytes(entity);
         }
         headers.setIfAbsent(Header.create(Header.CONTENT_LENGTH, String.valueOf(entityBytes.length)));
+        headers.setIfAbsent(USER_AGENT_HEADER);
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java:125]: nice

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: nima/http2/webclient/pom.xml [hunk 8]
```diff
@@ -54,6 +54,22 @@
             <artifactId>hamcrest-all</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.helidon.nima.http2</groupId>
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/webclient/pom.xml:58]: even with tests using `@ServerTest` ?

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/http1/ConnectionKey.java [hunk 9]
```diff
@@ -29,7 +30,7 @@
  * @param dnsResolver DNS resolver to be used
  * @param dnsAddressLookup DNS address lookup strategy
  */
-public record ConnectionKey(String scheme,
+record ConnectionKey(String scheme,
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/http1/ConnectionKey.java:33]: fixed

File: nima/http2/http2/src/main/java/io/helidon/nima/http2/Http2Util.java [hunk 10]
```diff
@@ -23,7 +23,11 @@
  * HTTP/2 utility.
  */
 public final class Http2Util {
-    private static final byte[] PRIOR_KNOWLEDGE_PREFACE =
+
+    /**
+     * Prior knowledge preface.
+     */
+    public static final byte[] PRIOR_KNOWLEDGE_PREFACE =
```
[the remark this answers is not in the document]
Comment (danielkec) [reply nima/http2/http2/src/main/java/io/helidon/nima/http2/Http2Util.java]: ugh

File: nima/http2/webclient/pom.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/http2/webclient/pom.xml:58]: Especially using `@ServerTest`, as it opens sockets and communicates over HTTP and network.
`@RoutingTest` can be a unit test.

[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/http2/webclient/pom.xml:58]: (Not yet supported with HTTP/2)

[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/http2/webclient/pom.xml:58]: You can use `@ServerTest`, but it should be configured as an integration test when used within this module

**Comment by danielkec** [comment]
Moved to PR #6399

**Pull request closed without merging**